processor/src/decoder/aux_trace/block_hash_table.rs

```diff
@@ -88,14 +89,11 @@ impl BlockHashTableRow {
 
     // Computes the initial row in the block hash table.
     pub fn table_init(main_trace: &MainTrace) -> Self {
-        let program_hash = {
-            let row_with_halt = main_trace
-                .row_iter()
-                .find(|&row| main_trace.get_op_code(row) == Felt::from(OPCODE_HALT))
-                .expect("execution trace must include at least one occurrence of HALT");
-
-            main_trace.decoder_hasher_state_first_half(row_with_halt)
-        };
+        let last_row = main_trace
+            .row_iter()
+            .nth(main_trace.num_rows() - NUM_RAND_ROWS - 1)
+            .expect("trace must have more than NUM_RAND_ROWS rows");
```

Comment:
This should work, but it feels like this logic belongs in the `MainTrace` - but it also would be a bit awkward to get it there (I think).

An alternative approach is to change the signature of this method to:
```rust
pub fn table_init(program_hash: Word) -> Self 
```
To do this, we'd also need to add `program_hash` field to [AuxTraceBuilder](https://github.com/0xPolygonMiden/miden-vm/blob/main/processor/src/decoder/aux_trace/mod.rs#L26), but that's shouldn't be too difficult since where `AuxTraceBuilder` is instantiated (i.e., in `Decoder::into_trace()`) we could get program hash as `self.program_hash()`.

Reply:
I went with the approach of feeding program_hash into the main_trace. The alternative involved refactoring a number of traits and didn't feel like an improvement, given that you mentioned that it feels like the program_hash should live inside main_trace. Happy to change again, LMK. Thanks

Reply:
Agreed that modifying `MainTrace` is probably a better idea. But if we are doing this, I'd probably add not `program_hash` to it, but the index of the last program row (i.e., the row at which the program execution ended and padding began - i.e., the row with the first `HALT` instruction). So, `MainTrace` could look something like this:
```rust
pub struct MainTrace {
    columns: ColMatrix<Felt>,
    last_program_row: RowIndex,
}
```

Then, the signature of `init_table()` could stay the same as before, and we'd use `last_program_row` to read the program hash from the main trace. What do you think?

btw, to get the index, I think we can use the `clk` variable from [here](https://github.com/0xPolygonMiden/miden-vm/blob/f5204afc4ea70d9295d7bd772ccb68f8e4d4fb39/processor/src/trace/mod.rs#L287).

Reply:
I've made those changes now. Wasn't sure whether to add the getter for last_program_row to main_trace or to add a program_hash() that uses the member variable instead. Went with the former approach